### ISSUE TITLE
Cleanup FUNCTOR: syntax

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1435]
If we make this:

```factor
USING: arrays assocs compiler.units eval fry interpolate
io.streams.string kernel lexer locals.parser locals.types make
multiline namespaces parser sequences vocabs.parser words ;

IN: functors2

SYNTAX: IS
    scan-token interpolate-locals
    '[ _ with-string-writer ] suffix! ;
    
<PRIVATE

: (parse-bindings) ( end -- words )
    [ dup parse-binding dup ]
    [ first2 [ make-local ] dip 2array ]
    produce 2nip ; 
    
: parse-bindings ( end -- words assoc )
    [
        building get use-words
        (parse-bindings)
    ] H{ } make ;

: parse-functor-body ( -- form )
    f [
        "WHERE" parse-bindings drop
        [ swap <def> suffix ] { } assoc>map concat
        ";FUNCTOR" parse-multiline-string interpolate-locals
        '[ @ _ with-string-writer parse-string drop ]
    ] with-lambda-scope ;

: (FUNCTOR2:) ( -- word def effect )
    scan-new-word [ parse-functor-body ] parse-locals-definition ;

PRIVATE>

SYNTAX: FUNCTOR2: (FUNCTOR2:) define-declared ;
```

Then we don't need all the ``FUNCTOR-SYNTAX:`` crap... only "problem" is that it looks a little uglier because it uses string interpolation...

This is the ``specialized-arrays``, and it works:

```
FUNCTOR2: define-array2 ( T -- )

A IS ${T}-array

WHERE

TUPLE: ${A}
{ underlying c-ptr read-only }
{ length array-capacity read-only } ; final

: <direct-${A}> ( alien len -- specialized-array ) ${A} boa ; inline

M: ${A} direct-like drop <direct-${A}> ; inline

: <${A}> ( n -- specialized-array )
    [ \ ${T} <underlying> ] keep <direct-${A}> ; inline

: (${A}) ( n -- specialized-array )
    [ \ ${T} (underlying) ] keep <direct-${A}> ; inline

M: ${A} clone [ underlying>> clone ] [ length>> ] bi <direct-${A}> ; inline

M: ${A} length length>> ; inline

M: ${A} nth-unsafe underlying>> \ ${T} alien-element ; inline

M: ${A} nth-c-ptr underlying>> \ ${T} array-accessor drop swap <displaced-alien> ; inline

M: ${A} set-nth-unsafe underlying>> \ ${T} set-alien-element ; inline

M: ${A} like drop dup A instance? [ \ T >c-array ] unless ; inline

M: ${A} new-sequence drop (${A}) ; inline

M: ${A} equal? over ${A} instance? [ sequence= ] [ 2drop f ] if ;

M: ${A} resize
    [
        [ \ ${T} heap-size * ] [ underlying>> ] bi*
        resize-byte-array
    ] [ drop ] 2bi
    <direct-${A}> ; inline

M: ${A} element-size drop \ ${T} heap-size ; inline

M: ${A} underlying-type drop \ ${T} ;

<< SYNTAX: ${A}{ \ } [ \ ${T} >c-array ] parse-literal ; >>

M: ${A} pprint-delims drop \ ${A}{ \ } ;

M: ${A} >pprint-sequence ;

INSTANCE: ${A} specialized-array

M: ${A} vs+ [ + \ ${T} c-type-clamp ] 2map ; inline
M: ${A} vs- [ - \ ${T} c-type-clamp ] 2map ; inline
M: ${A} vs* [ * \ ${T} c-type-clamp ] 2map ; inline

M: ${A} v*high [ * \ ${T} heap-size neg shift ] 2map ; inline

;FUNCTOR2
```

**Comment by erg** [comment]
The "modern" syntax proposal:

```
FUNCTOR: define-array ( T -- )

ALIAS: A          ${T}-array
ALIAS: <A>        <${A}>
ALIAS: (A)        (${A})
ALIAS: <direct-A> <direct-${A}>
ALIAS: A{         ${A}{
ALIAS: A@         ${A}@

TUPLE: A
{ underlying c-ptr read-only }
{ length array-capacity read-only } ; final

: <direct-A> ( alien len -- specialized-array ) A boa ; inline

M: A direct-like drop <direct-A> ; inline

: <A> ( n -- specialized-array )
    [ \ T <underlying> ] keep <direct-A> ; inline

: (A) ( n -- specialized-array )
    [ \ T (underlying) ] keep <direct-A> ; inline

M: A clone [ underlying>> clone ] [ length>> ] bi <direct-A> ; inline

M: A length length>> ; inline

M: A nth-unsafe underlying>> \ T alien-element ; inline

M: A nth-c-ptr underlying>> \ T array-accessor drop swap <displaced-alien> ; inline

M: A set-nth-unsafe underlying>> \ T set-alien-element ; inline

M: A like drop dup A instance? [ \ T >c-array ] unless ; inline

M: A new-sequence drop (A) ; inline

M: A equal? over A instance? [ sequence= ] [ 2drop f ] if ;

M: A resize
    [
        [ \ T heap-size * ] [ underlying>> ] bi*
        resize-byte-array
    ] [ drop ] 2bi
    <direct-A> ; inline

M: A element-size drop \ T heap-size ; inline

M: A underlying-type drop \ T ;

M: A pprint-delims drop \ A{ \ } ;

M: A >pprint-sequence ;

SYNTAX: A{ \ } [ \ T >c-array ] parse-literal ;

INSTANCE: A specialized-array

M: A vs+ [ + \ T c-type-clamp ] 2map ; inline
M: A vs- [ - \ T c-type-clamp ] 2map ; inline
M: A vs* [ * \ T c-type-clamp ] 2map ; inline

M: A v*high [ * \ T heap-size neg shift ] 2map ; inline

;FUNCTOR
```

**Comment by mrjbq7** [comment]
The problem with pure string interpolation (like the original suggestion above), is that it delays USING resolution and responsibility to the caller, which kinda sucks because it would be way better if the functor word resolved all the words that it depends on and the caller can just USE: the vocab with the functor, but not all the stuff it depended on.

**Comment by erg** [comment]
I don't think it has to be that way. ``FUNCTOR:`` could look up all the words that are used in definitions, and ``ALIAS:`` define the rest so that nothing is actually unresolved when the functor is parsed. It would be like ``ALIAS:`` words are deferred, and since you can't execute a functor directly it's fine.

**Comment by mrjbq7** [comment]
Nope it can't work that way... at least not until we have the modern parser that has no side effects... then the FUNCTOR could parse the "IS" aliases (supporting quotations, which are used in a couple of functors), then parse everything after the WHERE into a datastructure that is then "run" with the aliases in the qualified-words list when the function is run.  The parse should resolve all the USINGs so that the vocab that defines a FUNCTOR has all the dependencies and the caller only has to USE the vocab with the FUNCTOR that is being called.
